managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -2707,6 +2707,25 @@ public void operationFailed(ManagedLedgerException exception) {
         });
     }
 
+    /**
+     * Try set {@link #state} to {@link State#Closing}.
+     * @return false if the {@link #state} already is {@link State#Closing} or {@link State#Closed}.
+     */
+    private boolean trySetStateToClosing() {
+        final AtomicBoolean notClosing = new AtomicBoolean(false);
+        STATE_UPDATER.updateAndGet(this, state -> {
+            switch (state){
+                case Closing:
+                case Closed: return state;
+                default: {
+                    notClosing.set(true);
+                    return State.Closing;
+                }
```

Comment:
[`updateAndGet` doc](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicReferenceFieldUpdater.html#updateAndGet-T-java.util.function.UnaryOperator-) says:

> The function should be side-effect-free, since it may be re-applied when attempted updates fail due to contention among threads.

So I guess `notClosing.set(true)` may cause tricky bugs. How about just 
 comparing the state that before and after `updateAndGet`?

Reply:
Or, we can add `notClosing.set(false)` in `Closing` and `Closed` cases to make this operation idempotent.

Reply:
> Or, we can add notClosing.set(false) in Closing and Closed cases to make this operation idempotent.

Yes, you are right. already fixed